lib/prelude.dx

```diff
@@ -1066,6 +1067,127 @@ def float64_cosh(x:Float64) -> Float64 = %fdiv(%fadd(%exp(x), %exp(%fsub(f_to_f6
 def float64_tanh(x:Float64) -> Float64 = %fdiv(%fsub(%exp(x), %exp(%fsub(f_to_f64 0.0, x)))
                                                ,%fadd(%exp(x), %exp(%fsub(f_to_f64 0.0, x))))
 
+# Polynomial evaluation by Horner's method
+def unsafe_horner(x:a, ys:n=>a) -> a given (a|Add|Mul, n|Ix) =
```

Comment:
This duplicates the `evalpoly` function in the same file, though yours is maybe a little more general since it doesn't require a `zero`.

https://github.com/google-research/dex-lang/blob/main/lib/prelude.dx#L2225

Reply:
Ah right, thanks for pointing this out, I should have left a comment about it somewhere. The duplication is unfortunate but was intentional, because:
- `evalpoly`, nor any of the reduction machinery, isn't yet defined at this point in the prelude, and writing out a polynomial evaluation by Horner's method by hand with a lot of coefficients in multiple branches is quite tedious. (I believe one of the other special functions does this.) I was unsure about the ramifications of shuffling around the contents of the prelude to facilitate, so I left things in place.
- `evalpoly` requires a `VSpace` and `Float` but we want to be able to support both `Float` (AKA `Float32`) and `Float64`, but the latter doesn't implement `VSpace` (see also #1345).

I tried to differentiate it from `evalpoly` as more of an internal helper than an intended user-facing function in part by prefixing it with `unsafe_`, which seemed appropriate since it directly uses two functions similarly marked as unsafe.

> though yours is maybe a little more general since it doesn't require a `zero`.

As an aside, IIUC, I think mine does still require a `zero` since that's part of the `Add` interface and I have `a|Add`.

Reply:
Oh, thanks for explaining.  Yes, moving things within the prelude is a punishing game of dependency tetris.  @dougalm has usually been fine with me moving things around as long as the tests still pass, but I would be surprised if you could move `evalpoly` up 1000 lines and have it still work.

Good point about the `zero`, I got mixed up about which method is in which class.

Reply:
Actually, it might take a bit of additional shuffling and/or finagling, but I think the `Floating` interface and its instances could be moved down below `evalpoly`. That way they'd also be able to use `copysign`, `inf`, and `nan` (at least for the `Float32` implementation, anyway). I believe the only use of `Floating` prior to the definition of `evalpoly` is for `std`, which would need to be moved as well. Though even if this shuffling is done, `evalpoly` itself would need to be amended in order to be usable for the `Float64` definition.